File: backend/internal/service/group_realtime.go
package service

import (
	"context"
	"sync"
	"time"

	infraerrors "github.com/Wei-Shaw/sub2api/internal/pkg/errors"
)

// GroupRealtimeStore reads existing PostgreSQL logs; requests do not write monitoring counters.
type GroupRealtimeStore interface {
	Snapshot(context.Context) (*GroupRealtimeSnapshot, error)
}

type GroupRealtimeRow struct {
	GroupID     int64            `json:"group_id"`
	GroupName   string           `json:"group_name"`
	Platform    string           `json:"platform"`
	Status      string           `json:"status"`
	RPM         int64            `json:"rpm"`
	Success     int64            `json:"success"`
	Failed      int64            `json:"failed"`
	Cancelled   int64            `json:"cancelled"`
	SuccessRate *float64         `json:"success_rate"`
	Failures    map[string]int64 `json:"failures"`
}

type GroupRealtimeSnapshot struct {
	StartTime       time.Time          `json:"start_time"`
	EndTime         time.Time          `json:"end_time"`
	CollectingSince time.Time          `json:"collecting_since"`
	Partial         bool               `json:"partial"`
	Groups          []GroupRealtimeRow `json:"groups"`
}

type GroupRealtimeMonitor struct {
	store          GroupRealtimeStore
	cacheMu        sync.Mutex
	cached         *GroupRealtimeSnapshot
	cachedAt       time.Time
	historyMu      sync.Mutex
	history        *GroupRealtimeHistory
	historyMinutes int
	historyAt      time.Time
}

func NewGroupRealtimeMonitor(store GroupRealtimeStore) *GroupRealtimeMonitor {
	return &GroupRealtimeMonitor{store: store}
}

func (s *OpsService) GetGroupRealtime(ctx context.Context) (*GroupRealtimeSnapshot, error) {
	if err := s.RequireMonitoringEnabled(ctx); err != nil {
		return nil, err
	}
	if s.groupRealtime == nil || s.groupRealtime.store == nil {
		return nil, infraerrors.ServiceUnavailable("GROUP_REALTIME_UNAVAILABLE", "Group realtime collector unavailable")
	}
	m := s.groupRealtime
	m.cacheMu.Lock()
	defer m.cacheMu.Unlock()
	if m.cached != nil && time.Since(m.cachedAt) < 3*time.Second {
		return m.cached, nil
	}
	result, err := m.store.Snapshot(ctx)
	if err != nil {
		return nil, err
	}
	m.cached, m.cachedAt = result, time.Now()
	return result, nil
}

File: frontend/src/api/admin/groupRealtime.ts
import { apiClient } from '../client'

export interface GroupRealtimeRow {
  group_id: number
  group_name: string
  platform: string
  status: string
  rpm: number
  success: number
  failed: number
  cancelled: number
  success_rate: number | null
  failures: Record<string, number>
}

export interface GroupRealtimeSnapshot {
  start_time: string
  end_time: string
  collecting_since: string
  partial: boolean
  groups: GroupRealtimeRow[]
}

export async function getGroupRealtime(signal?: AbortSignal): Promise<GroupRealtimeSnapshot> {
  const { data } = await apiClient.get<GroupRealtimeSnapshot>('/admin/ops/group-realtime', { signal })
  return data
}

export interface GroupHistoryRow {
  points: GroupHistoryPoint[]
  group_id: number
  group_name: string
  platform: string
  status: string
  started: number
  success: number
  failed: number
  cancelled: number
  rpm: number | null
  success_rate: number | null
}
export interface GroupHistoryPoint {
  bucket_start: string
  rpm: number | null
  success_rate: number | null
  started: number
  success: number
  failed: number
  cancelled: number
  partial: boolean
}
export interface GroupHistory {
  start_time: string
  end_time: string
  collecting_since: string
  bucket_seconds: number
  covered_seconds: number
  partial: boolean
  group_id: number | null
  groups: GroupHistoryRow[]
  points: GroupHistoryPoint[]
}
export type GroupHistoryWindow = 15 | 60 | 360 | 1440 | 10080
export async function getGroupHistory(windowMinutes: GroupHistoryWindow, groupID: number | null, signal?: AbortSignal): Promise<GroupHistory> {
  const { data } = await apiClient.get<GroupHistory>('/admin/ops/group-realtime/history', {
    params: { window_minutes: windowMinutes, ...(groupID !== null ? { group_id: groupID } : {}) }, signal,
  })
  return data
}
